### PR TITLE
Enhancements for `future` and `shared_future`

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
This PR is making some enhancements for `<future>`.

1. Currently `std::is_nothrow_move_assignable_v<std::shared_future<void>>` is `false` in MSVC STL ([Godbolt link](https://godbolt.org/z/fvb85K8Wv)), which is non-conforming. This PR fixes this bug by making internal functions `noexcept` (inspired by #3278) and some special member functions defaulted.
2. Currently users can touch internal constructors by definition `std::future<T> ft(another_future, {});` (where `another_future` is of type `std::future<T>` or `std::shared_future<T>`) ([Godbolt link](https://godbolt.org/z/eqvr8nzPx)). This PR forbids such definitions - because I think the internal constructors shouldn't be directly touched when no _Ugly name is explicitly specified, although the status quo is conforming due to [[member.functions]/2](https://eel.is/c++draft/member.functions#2).
3. Updating the references to refer to WG21-N4917 and dropping the section number. Towards #182.
4. (Doing while reviewing) Manually inlining `_Copy_from` and `_Move_from` of `_State_manager` into constructors and assignment operators, and simply the codes - this is making the implementation code shorter.

### CODE REVIEW

File: stl/inc/future [hunk 1]
```diff
@@ -927,18 +924,13 @@ private:
     using _Mybase = _State_manager<_Ty*>;
 
 public:
-    future() noexcept {}
+    future() = default;
```
Comment (strega-nil-ms) [inline stl/inc/future:927]: This is a terrifying change; however, I _do_ think it's fine, since `future` will never be trivially default constructible..? Same for the destructor.

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
This generally seems like a good change to me; there are a few extra `noexcept`s that are needed, imo, so that the compiler doesn't generate `try`s.

File: stl/inc/future (hunk 1 above)
Comment (StephanTLavavej) [reply stl/inc/future:927]: Changing copy ctor triviality affects ABI (as infamously encountered in `tuple<>`) - but that doesn't apply here, so I think you're correct.

**Review by strega-nil-ms — APPROVED** [review]
LGTM, thanks so much!

File: stl/inc/future [hunk 2]
```diff
@@ -170,7 +172,7 @@ class _Associated_state;
 
 template <class _Ty>
 struct __declspec(novtable) _Deleter_base { // abstract base class for managing deletion of state objects
-    virtual void _Delete(_Associated_state<_Ty>*) = 0;
+    virtual void _Delete(_Associated_state<_Ty>*) noexcept = 0;
     virtual ~_Deleter_base() noexcept {}
```
Comment (CaseyCarter) [inline stl/inc/future:176]: While we're here, we may as well:
```suggestion
    virtual ~_Deleter_base() = default;
```

File: stl/inc/future [hunk 3]
```diff
@@ -725,21 +727,20 @@ template <class _Ty>
 class _State_manager {
     // class for managing possibly non-existent associated asynchronous state object
 public:
-    _State_manager() : _Assoc_state(nullptr) { // construct with no associated asynchronous state object
-        _Get_only_once = false;
+    _State_manager() noexcept
+        : _Assoc_state(nullptr), _Get_only_once(false) { // construct with no associated asynchronous state object
     }
```
Comment (CaseyCarter) [inline stl/inc/future:732]: No change requested, but I note that in our modern coding style we'd make these default member initializers and `= default` this constructor (letting the compiler make it implicitly `noexcept`). (This would let us clean up a lot of `_Assoc_state(nullptr)` mem-initializers.)

File: stl/inc/future [hunk 4]
```diff
@@ -725,21 +727,20 @@ template <class _Ty>
...
+    _State_manager() noexcept
+        : _Assoc_state(nullptr), _Get_only_once(false) { // construct with no associated asynchronous state object
     }
 
-    _State_manager(_Associated_state<_Ty>* _New_state, bool _Get_once)
-        : _Assoc_state(_New_state) { // construct with _New_state
-        _Get_only_once = _Get_once;
+    _State_manager(_Associated_state<_Ty>* _New_state, bool _Get_once) noexcept
+        : _Assoc_state(_New_state), _Get_only_once(_Get_once) { // construct with _New_state
     }
```
Comment (CaseyCarter) [inline stl/inc/future:736]:
```suggestion
        : _Assoc_state(_New_state), _Get_only_once(_Get_once) {} // construct with _New_state
```

File: stl/inc/future [hunk 5]
```diff
@@ -831,15 +832,16 @@ public:
...
 
-    _Associated_state<_Ty>* _Ptr() const {
+    _Associated_state<_Ty>* _Ptr() const noexcept {
         return _Assoc_state;
     }
 
-    void _Copy_from(const _State_manager& _Other) { // copy stored associated asynchronous state object from _Other
+    void _Copy_from(
+        const _State_manager& _Other) noexcept { // copy stored associated asynchronous state object from _Other
```
Comment (CaseyCarter) [inline stl/inc/future:844]:
```suggestion
    void _Copy_from(const _State_manager& _Other) noexcept {
        // copy stored associated asynchronous state object from _Other
```

File: tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp [hunk 6]
```diff
@@ -260,12 +273,45 @@ void test_shared_future_noexcept_copy_impl() {
...
+template <typename T, typename... Args>
+constexpr bool
+    is_constructible_with_trailing_empty_brace_impl<void_t<decltype(T(declval<Args>()..., {}))>, T, Args...> = true;
+
+template <typename T, typename... Args>
+constexpr bool is_constructible_with_trailing_empty_brace =
+    is_constructible_with_trailing_empty_brace_impl<void, T, Args...>;
+
+STATIC_ASSERT(is_constructible_with_trailing_empty_brace<pair<int, int>, int>);
+STATIC_ASSERT(is_constructible_with_trailing_empty_brace<pair<int, int>, const int&>);
```
Comment (CaseyCarter) [inline tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp:299]: This is handy to prove to yourself that the trait you wrote works, but probably not necessary in the long term. (We conventionally only write tests for very complicated test machinery.)
```suggestion
```

File: tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp [hunk 7]
```diff
@@ -260,12 +273,45 @@ void test_shared_future_noexcept_copy_impl() {
...
+    test_future_shared_future_noexcept_impl<int>();
+    test_future_shared_future_noexcept_impl<int&>();
+    test_future_shared_future_noexcept_impl<void>();
+
     test_shared_future_noexcept_copy_impl<int>();
     test_shared_future_noexcept_copy_impl<int&>();
     test_shared_future_noexcept_copy_impl<void>();
 }
 
+// Also test the non-constructibility of future from (future {}) and (shared_future, {})
```
Comment (CaseyCarter) [inline tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp:286]:
```suggestion
// Also test the non-constructibility of future from (future, {}) and (shared_future, {})
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/future (hunk 3 above)
Comment (frederick-vs-ja) [reply stl/inc/future]: I considered default member initializers for this but gave up the changes, because `_Get_only_once` might be writen 3 times in the copy/move ctors. Perhaps this should be done in some further refactoring.

File: stl/inc/future (hunk 2 above)
Comment (frederick-vs-ja) [reply stl/inc/future]: I'm not sure whether it would be better to make this dtor non-virtual (and implicitly declared), because this dtor seems never virtually called. Perhaps such change should be done in vNext?

File: stl/inc/future (hunk 3 above)
Comment (frederick-vs-ja) [reply stl/inc/future]: Now I think such further refactoring is worthy and trying to do it.

File: stl/inc/future (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/future]: `protected` and non-`virtual` _would_ be better, but IMO has no benefits commensurate with the risks of ABI issues, especially in CLR metadata. We could add a `// TRANSITION, vNext (make non-virtual, never called polymorphically)` and/or add a bullet to the "things to change in vNext" issue.

File: stl/inc/future [hunk 8]
```diff
@@ -173,7 +173,7 @@ class _Associated_state;
 template <class _Ty>
 struct __declspec(novtable) _Deleter_base { // abstract base class for managing deletion of state objects
     virtual void _Delete(_Associated_state<_Ty>*) noexcept = 0;
-    virtual ~_Deleter_base() noexcept {}
+    virtual ~_Deleter_base()                               = default;
```
Comment (CaseyCarter) [inline stl/inc/future:176]: Great work, clang-format. 🙄 (No change requested.)

File: stl/inc/future [hunk 9]
```diff
@@ -752,12 +749,34 @@ public:
...
     _State_manager& operator=(_State_manager&& _Other) noexcept {
-        _Move_from(_Other);
+        // move stored associated asynchronous state object from _Other
+        if (this != _STD addressof(_Other)) {
+            if (_Assoc_state) {
+                _Assoc_state->_Release();
+            }
+
+            _Assoc_state        = _Other._Assoc_state;
+            _Other._Assoc_state = nullptr;
```
Comment (CaseyCarter) [inline stl/inc/future:777]: For consistency with the constructor:
```suggestion
            _Assoc_state        = _STD exchange(_Other._Assoc_state, nullptr);
```

File: stl/inc/future [hunk 10]
```diff
@@ -752,12 +749,34 @@ public:
...
+            }
+
+            if (_Other._Assoc_state) { // do the copy
+                _Other._Assoc_state->_Retain();
+                _Assoc_state   = _Other._Assoc_state;
+                _Get_only_once = _Other._Get_only_once;
+            } else {
+                _Assoc_state = nullptr;
+            }
+        }
```
Comment (CaseyCarter) [inline stl/inc/future:765]: I _hate_ self-checks in assignment operators, they pessimize the common case. We can get rid of this one with the "increment the reference count before decrementing" pattern:
```suggestion
        if (_Other._Assoc_state) {
            _Other._Assoc_state->_Retain();
        }

        if (_Assoc_state) {
            _Assoc_state->_Release();
        }

        _Assoc_state   = _Other._Assoc_state;
        _Get_only_once = _Other._Get_only_once;
```

**Review by CaseyCarter — APPROVED** [review]
These are suggestions rather than requirements, so I'll go ahead and approve.

File: stl/inc/future (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/future]: I just saw https://github.com/microsoft/STL/issues/208#issuecomment-1350452371 which records this suggestion, so I think my "add a comment" suggestion is unnecessary.

File: stl/inc/future (hunk 8 above)
Comment (frederick-vs-ja) [reply stl/inc/future]: Fortunately, clang-format can be tweaked by the long comment you suggested.😎

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/future (hunk 8 above)
Comment (CaseyCarter) [reply stl/inc/future]: Oooh, nice!

File: stl/inc/future (hunk 10 above)
Comment (frederick-vs-ja) [reply stl/inc/future]: This may have negative impact of performance on self-copy-assignment, though...

Comment (CaseyCarter) [reply stl/inc/future]: I'm happy to pessimize a rare corner case to reduce total codesize and speedup the common case.

File: tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp [hunk 11]
```diff
@@ -260,12 +273,42 @@ void test_shared_future_noexcept_copy_impl() {
...
     test_shared_future_noexcept_copy_impl<void>();
 }
 
+// Also test the non-constructibility of future from (future, {}) and (shared_future, {})
+template <typename Void, typename T, typename... Args>
+constexpr bool is_constructible_with_trailing_empty_brace_impl = false;
+
+template <typename T, typename... Args>
+constexpr bool
+    is_constructible_with_trailing_empty_brace_impl<void_t<decltype(T(declval<Args>()..., {}))>, T, Args...> = true;
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp:292]: Test nitpick: When implementing custom type traits, it's a good idea for the test to have at least one positive and one negative case. This prevents the type trait itself from being damaged in a way that always returns `true` or `false`. Here, all of the usage is verifying `!is_constructible_with_trailing_empty_brace`, so we should add a positive case - `pair` would suffice (it actually has `<class U1 = T1, class U2 = T2>` specifically to handle empty braces).

**Comment by StephanTLavavej** [comment]
Thanks - this is riskier than the average PR, but I think it preserves ABI and is a worthy improvement.

I've pushed a couple of tiny changes to the test - a trivial naming nitpick (which I wouldn't have bothered with by itself), and a line to verify that the empty-braces type trait can report both `true` and `false`, since that's a way for test coverage to be invisibly damaged. FYI @CaseyCarter @strega-nil-ms.

File: tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp [hunk 12]
```diff
@@ -295,6 +295,8 @@ template <typename T, typename... Args>
 constexpr bool is_constructible_with_trailing_empty_brace =
     is_constructible_with_trailing_empty_brace_impl<void, T, Args...>;
 
+STATIC_ASSERT(is_constructible_with_trailing_empty_brace<pair<double*, int>, double*>); // verify a true case
```
Comment (CaseyCarter) [inline tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp:298]: Me: "We don't need tests for such simple test machinery - you can remove these."
Stephan: Adds a test for the test machinery.

**Review by CaseyCarter — APPROVED** [review]

File: tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp (hunk 12 above)
Comment (StephanTLavavej) [reply tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp:298]: It's partially a general principle about being methodical with testing, but also it's paranoia about compiler bugs. One never knows whether a compiler today or tomorrow will have trouble with this brace SFINAE and begin reporting `false` all the time. On a related note, I found several cases when working on Standard Library Modules where very simple tests that "obviously" passed for years, were just special enough to reveal bugs when the compiler began working totally differently.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

File: tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp (hunk 12 above)
Comment (CaseyCarter) [reply tests/std/tests/Dev11_0235721_async_and_packaged_task/test.cpp:298]: We had a wonderful discussion in the LWG telecon this morning about how much easier library development would be if it weren't for compilers.

**Comment by StephanTLavavej** [comment]
Thanks for noticing the nonconformance here and carefully improving things! :detective: :heart_eyes_cat: :sparkles:
